pyi.py

```diff
@@ -708,6 +682,96 @@ def _analyse_typing_Literal(node: ast.Subscript) -> TypingLiteralAnalysis:
     )
 
 
+_KNOWN_ENUM_BASES = frozenset(
+    {"Enum", "Flag", "IntEnum", "IntFlag", "StrEnum", "ReprEnum"}
+)
+
+
+_COMMON_METACLASSES = {
+    "type": "builtins",
+    "ABCMeta": "abc",
+    "EnumMeta": "enum",
+    "EnumType": "enum",
+}
+
+
+@dataclass(frozen=True)
+class EnclosingClassContext:
+    node: ast.ClassDef
+    cls_name: str
+    bases_map: defaultdict[str, set[str | None]]
+
+    def contains_in_bases(self, obj: str, *, from_: AbstractSet[str]) -> bool:
+        if obj not in self.bases_map:
+            return False
+        if None in self.bases_map[obj]:
+            return True
+        return bool(self.bases_map[obj] & from_)
+
+    @cached_property
+    def is_protocol_class(self) -> bool:
+        return self.contains_in_bases("Protocol", from_=_TYPING_MODULES)
+
+    @cached_property
+    def is_typeddict_class(self) -> bool:
+        return self.contains_in_bases(
+            "TypedDict", from_=(_TYPING_MODULES | {"mypy_extensions"})
+        )
+
+    @cached_property
+    def is_enum_class(self) -> bool:
+        return any(
+            self.contains_in_bases(enum_cls, from_={"enum"})
+            for enum_cls in _KNOWN_ENUM_BASES
+        )
+
+    @cached_property
+    def is_metaclass(self) -> bool:
+        return any(
+            self.contains_in_bases(metacls, from_={module})
+            for metacls, module in _COMMON_METACLASSES.items()
+        )
+
+    @cached_property
+    def is_decorated_with_final(self) -> bool:
+        return any(_is_final(decorator) for decorator in self.node.decorator_list)
+
+
+class ClassBase(NamedTuple):
+    module: str | None
+    obj: str
+
+
+def _analyze_classdef(node: ast.ClassDef) -> EnclosingClassContext:
+    bases_map: defaultdict[str, set[str | None]] = defaultdict(set)
+
+    def _unravel(node: ast.expr) -> str | None:
+        if isinstance(node, ast.Name):
+            return node.id
+        if isinstance(node, ast.Attribute):
+            return f"{_unravel(node.value)}.{node.attr}"
+        return None
```

Comment:
Can this happen with the current ast? Wouldn't it be safer to raise an exception here, or use an assertion?

Reply:
It _can_ definitely happen; the `.value` of an `Attribute` node can be any number of things:

```pycon
>>> class EvenStranger:
...     def __getitem__(self, key):
...         return self
...     def __getattr__(self, attr):
...         return int
...
>>> class Foo(EvenStranger()[1][2].foo): ...
>>> tree = ast.parse("EvenStranger()[1][2].foo")
>>> print(ast.dump(tree, indent=2))
Module(
  body=[
    Expr(
      value=Attribute(
        value=Subscript(
          value=Subscript(
            value=Call(
              func=Name(id='EvenStranger', ctx=Load()),
              args=[],
              keywords=[]),
            slice=Constant(value=1),
            ctx=Load()),
          slice=Constant(value=2),
          ctx=Load()),
        attr='foo',
        ctx=Load()))],
  type_ignores=[])
```

This is obviously pretty unlikely to happen in a stub, but I don't think we should be raising an exception if we come across something like this